sympy/tensor/array/expressions/__init__.py

```diff
@@ -0,0 +1,108 @@
+r"""
+Array expressions are expressions representing N-dimensional arrays, without evaluating them.
+They are *lazy* as they don't evaluate the operations on the arrays until required and
+can define operations on unknown array (in a certain way reminiscent of graphs of
+operations in TensorFlow).
+
+Every N-dimensional array operator has a corresponding lazy object.
+
+Table of correspondences:
+
+=============================== =============================
+         Array operator           Array expression operator
+=============================== =============================
+        tensorproduct                 ArrayTensorProduct
+        tensorcontraction             ArrayContraction
+        tensordiagonal                ArrayDiagonal
+        permutedims                   PermuteDims
+        derive_by_array               array_derive
```

Comment:
It is confusing to me to call `array_derive` a "corresponding lazy object" here and as far as I can see, it isn't lazy. e.g. when dispatched an `ArraySymbol` it will yield zero's or an identity with some shape depending on whether we derive by the same object or not. Lazy would imply to me that we could still substitute either of the objects later on. (To indicate that they are equal after all) e.g.
```python
from sympy import Derivative
from sympy.abc import x,y
Derivative(x,y)
Derivative(x,y).subs(y,x)
```

Reply:
Well, I have to remove the word *lazy* from the documentation, it's the wrong adjective. Better call it *array expression operator* or *component-implicit* (whereas *component-explicit* are traditional arrays with elements).